File: java/src/test/be/steinsomers/bron_kerbosch/DegeneracyFilterTest.java
package be.steinsomers.bron_kerbosch;

import net.jqwik.api.Arbitraries;
import net.jqwik.api.Arbitrary;
import net.jqwik.api.ForAll;
import net.jqwik.api.Property;
import net.jqwik.api.Provide;
import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.SortedSet;
import java.util.TreeSet;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

final class DegeneracyFilterTest {
    private static SortedSet<Integer> sortedDegeneracyFilterIncludingNeighbours(final UndirectedGraph g) {
        final SortedSet<Integer> vertices = new TreeSet<>();
        new DegeneracyFilter(g).forEachRemaining((final int v) -> {
            vertices.add(v);
            vertices.addAll(g.neighbours(v));
        });
        return vertices;
    }

    @Test
    void empty() {
        final var g = new UndirectedGraph(List.of());
        var f = new DegeneracyFilter(g);
        assertFalse(f.hasNext());
    }

    @Test
    void pair() {
        final var g = new UndirectedGraph(List.of(Set.of(1), Set.of(0)));
        var f = new DegeneracyFilter(g);
        assertTrue(f.hasNext());
        f.nextInt();
        assertFalse(f.hasNext());
    }

    @Test
    void split() {
        final var g = new UndirectedGraph(List.of(Set.of(1), Set.of(0, 2), Set.of(1)));
        var f = new DegeneracyFilter(g);
        assertTrue(f.hasNext());
        var first = f.nextInt();
        assertNotEquals(1, first);
        assertTrue(f.hasNext());
        var second = f.nextInt();
        assertNotEquals(first, second);
        assertFalse(f.hasNext());
    }

    /**
     * @param adjacencyLikes List of suggested neighbours, indexed by vertex. The list is oblivious
     *                       to symmetry. If a vertex appears as its own neighbour, that entry will
     *                       be ignored. The list may be empty to begin with. The latter two
     *                       properties make it likely a vertex is unconnected, but the need for
     *                       symmetry makes it likely that another vertex connects to it anyway.
     */
    private static List<Set<Integer>> makeSymmetricAdjacencies(final List<Set<Integer>> adjacencyLikes) {
        final var order = adjacencyLikes.size();
        final List<Set<Integer>> adjacencies = Stream
                .generate(() -> (Set<Integer>) new HashSet<Integer>(16))
                .limit(order)
                .toList();
        for (int v = 0; v < order; ++v) {
            final var neighbours = adjacencyLikes.get(v);
            for (final int w : neighbours) {
                if (v < w) {
                    adjacencies.get(v).add(w);
                    adjacencies.get(w).add(v);
                }
            }
        }
        return adjacencies;
    }

    @Property
    boolean degeneracyFilterCoversConnectedVertices(
            @ForAll("arbitraryAdjacencyLikes") final List<Set<Integer>> adjacencyLikes) {
        final var adjacencies = makeSymmetricAdjacencies(adjacencyLikes);
        final var g = new UndirectedGraph(adjacencies);
        final SortedSet<Integer> connectedVertices =
                g.connectedVertices().collect(Collectors.toCollection(TreeSet::new));
        return sortedDegeneracyFilterIncludingNeighbours(g).equals(connectedVertices);
    }

    @Property
    boolean degeneracyFilterDropsSome(
            @ForAll("arbitraryAdjacencyLikes") final List<Set<Integer>> adjacencyLikes) {
        final var adjacencies = makeSymmetricAdjacencies(adjacencyLikes);
        final var g = new UndirectedGraph(adjacencies);
        final var connected = g.connectedVertices().count();
        final var filtered = new DegeneracyFilter(g).stream().count();
        return connected == 0 ? filtered == connected : filtered < connected;
    }

    @Property
    boolean degeneracyFilterStartsWithLowestDegree(
            @ForAll("arbitraryAdjacencyLikes") final List<Set<Integer>> adjacencyLikes) {
        final var adjacencies = makeSymmetricAdjacencies(adjacencyLikes);
        final var g = new UndirectedGraph(adjacencies);
        final var f = new DegeneracyFilter(g);
        if (f.hasNext()) {
            final int first = f.nextInt();
            return f.stream().allMatch(v -> g.degree(first) <= g.degree(v));
        } else {
            return true;
        }
    }

    private static Arbitrary<Set<Integer>> arbitraryNeighbours(final int order) {
        return Arbitraries.integers().between(0, order - 1).set();
    }

    // Provide arbitrary input for makeSymmetricAdjacencies.
    @Provide
    static Arbitrary<List<Set<Integer>>> arbitraryAdjacencyLikes() {
        final Arbitrary<Integer> order = Arbitraries.integers().between(1, 99);
        return order.flatMap(o -> arbitraryNeighbours(o).list().ofSize(o));
    }
}

File: kotlin/src/test/be/steinsomers/bron_kerbosch/DegeneracyFilterTest.kt
package be.steinsomers.bron_kerbosch

import net.jqwik.api.*
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.*
import java.util.function.IntConsumer
import java.util.stream.Stream

internal class DegeneracyFilterTest {
    private fun sortedDegeneracyOrderingIncludingNeighbours(g: UndirectedGraph): SortedSet<Int> {
        val vertices: SortedSet<Int> = TreeSet()
        DegeneracyFilter(g).forEachRemaining(IntConsumer { v: Int ->
            vertices.add(v)
            vertices.addAll(g.neighbours(v))
        })
        return vertices
    }

    /**
     * @param adjacencyLikes List of suggested neighbours, indexed by vertex. The list is oblivious
     * to symmetry. If a vertex appears as its own neighbour, that entry will
     * be ignored. The list may be empty to begin with. The latter two
     * properties make it likely a vertex is unconnected, but the need for
     * symmetry makes it likely that another vertex connects to it anyway.
     */
    private fun makeSymmetricAdjacencies(adjacencyLikes: List<Set<Int>>): List<Set<Int>> {
        val order = adjacencyLikes.size
        val adjacencies = Stream
            .generate { HashSet<Int>(16) as MutableSet<Int> }
            .limit(order.toLong())
            .toList()
        for (v in 0..<order) {
            val neighbours = adjacencyLikes[v]
            for (w in neighbours) {
                if (v < w) {
                    adjacencies[v].add(w)
                    adjacencies[w].add(v)
                }
            }
        }
        return adjacencies
    }

    @Test
    fun empty() {
        val g = UndirectedGraph(listOf())
        val f = DegeneracyFilter(g)
        Assertions.assertFalse(f.hasNext())
    }

    @Test
    fun pair() {
        val g = UndirectedGraph(listOf(setOf(1), setOf(0)))
        val f = DegeneracyFilter(g)
        Assertions.assertTrue(f.hasNext())
        f.nextInt()
        Assertions.assertFalse(f.hasNext())
    }

    @Test
    fun split() {
        val g = UndirectedGraph(listOf(setOf(1), setOf(0, 2), setOf(1)))
        val f = DegeneracyFilter(g)
        Assertions.assertTrue(f.hasNext())
        val first = f.nextInt()
        Assertions.assertNotEquals(1, first)
        Assertions.assertTrue(f.hasNext())
        val second = f.nextInt()
        Assertions.assertNotEquals(first, second)
        Assertions.assertFalse(f.hasNext())
    }

    @Property
    fun degeneracyOrderingCoversConnectedVertices(
        @ForAll("arbitraryAdjacencyLikes") adjacencyLikes: List<Set<Int>>
    ): Boolean {
        val adjacencies = makeSymmetricAdjacencies(adjacencyLikes)
        val g = UndirectedGraph(adjacencies)
        val connectedVertices: SortedSet<Int> = g.connectedVertices(TreeSet())
        return sortedDegeneracyOrderingIncludingNeighbours(g) == connectedVertices
    }

    @Property
    fun degeneracyOrderingDropsSome(
        @ForAll("arbitraryAdjacencyLikes") adjacencyLikes: List<Set<Int>>
    ): Boolean {
        val adjacencies: List<Set<Int>> = makeSymmetricAdjacencies(adjacencyLikes)
        val g = UndirectedGraph(adjacencies)
        val connectedVertices = g.connectedVertices(HashSet())
        val filtered = DegeneracyFilter(g).stream().count()
        return if (connectedVertices.isEmpty()) {
            filtered == 0L
        } else {
            filtered < connectedVertices.size
        }
    }

    @Property
    fun degeneracyOrderingStartsWithLowestDegree(
        @ForAll("arbitraryAdjacencyLikes") adjacencyLikes: List<Set<Int>>
    ): Boolean {
        val adjacencies = makeSymmetricAdjacencies(adjacencyLikes)
        val g = UndirectedGraph(adjacencies)
        val f = DegeneracyFilter(g)
        return if (f.hasNext()) {
            val first = f.nextInt()
            f.stream().allMatch { v -> g.degree(first) <= g.degree(v) }
        } else {
            true
        }
    }

    private fun arbitraryNeighbours(order: Int): Arbitrary<Set<Int>> {
        return Arbitraries.integers().between(0, order - 1).set()
    }

    // Provide arbitrary input for makeSymmetricAdjacencies.
    @Suppress("unused")
    @Provide
    private fun arbitraryAdjacencyLikes(): Arbitrary<List<Set<Int>>> {
        val order: Arbitrary<Int> = Arbitraries.integers().between(1, 99)
        return order.flatMap { o: Int -> arbitraryNeighbours(o).list().ofSize(o) }
    }
}
